### ISSUE TITLE
Split SDK into feature-based modules

### ISSUE THREAD

**Issue by armcknight** [getsentry/sentry-cocoa#3219]
### Description

Customers should be able to selectively integrate separate SDK modules depending on the features they'd like to use. This would optimize the SDK footprint in each app, as instead of still shipping binary for unused features, they wouldn't be present at all.

This has other knock on benefits, like improved build times and enforcing cleaner APIs.

**Comment by bruno-garcia** [comment]
How does this affect installation steps?

**Comment by armcknight** [comment]
We can keep the current package as one that includes everything, and then have subpackages like sentry-core (which would do error/crash reporting), sentry-performance and sentry-UIKit, among other things, that could all be automatically included in the top-level Sentry package. So it shouldn't break anybody.

**Comment by bruno-garcia** [comment]
That makes sense. We did that on Android.
`sentry-android` is a meta-package that includes `sentry-android-core` (actualy has the Android specific core), `sentry-android-ndk` (native stuff) and `sentry-java` the core Java-only which is used also by backend Java/Kotlin/JVM.

So users who don't want the additional 1MB or whatever NDK support, could go straight to `sentry-android-core` but that's rare tbqh. We did need to have two packages for Android and Java (reusable) which is the reason we ended up with this design. All while allowing a simple installation method: Add `sentry-android`.

**Comment by armcknight** [comment]
It's soundung more like this is going to be needed for the long term health of the package. We more recently looked at SDK size, and splitting it up can really reduce the footprint for the average customer who may not want every single feature we offer. It would also help with the ongoing challenges around a UIKitless offering.

**Comment by philprime** [comment]
cc @noahsmartin @philipphofmann @itaybre based on our most recent discussion regarding a project restructuring, should we close this issue as "Not planned"? IIRC you brought up that you advocated that we should not do this.

**Comment by philippsuess** [comment]
Yes, let's close this and open up new issues once we know exactly what we plan on doing. Please reopen, if you disagree.
